lib/mongodb.js

```diff
@@ -1870,30 +1873,71 @@ MongoDB.prototype.ping = function(cb) {
   }
 };
 
+// Determine if a property must be stored as ObjectID
+function mustBeObjectID(propDef) {
+  if (!propDef) return false;
+  if (propDef.mongodb && ObjectIdTypeRegex.test(propDef.mongodb.dataType)) return true;
+  return false;
+}
+
+// Determine if strictObjectIDCoercion should be enabled
+function isStrictObjectIDCoercionEnabled(modelCtor, options) {
+  const settings = modelCtor.settings;
+  return (settings && settings.strictObjectIDCoercion) ||
+    (modelCtor.model && modelCtor.model.getConnector().settings.strictObjectIDCoercion) ||
+    options &&
+    options.strictObjectIDCoercion;
+}
+
+function coerceToObjectId(modelCtor, propDef, propValue) {
+  if (mustBeObjectID(propDef)) {
+    if (isObjectIDProperty(modelCtor, propDef, propValue)) {
+      return ObjectID(propValue);
+    } else {
+      throw new Error(`${propValue} is not an ObjectID string`);
+    }
+  } else if (isStrictObjectIDCoercionEnabled(modelCtor)) {
+    if (isObjectIDProperty(modelCtor, propDef, propValue)) {
+      return ObjectID(propValue);
+    }
+  } else if (ObjectIdValueRegex.test(propValue)) {
+    return ObjectID(propValue);
```

Comment:
Here we don't need to test `if (isObjectIDProperty(modelCtor, propDef, propValue))`?

Reply:
`ObjectIdValueRegex.test()` is checking if the string looks like an ObjectID.

In other places, we do `isObjectIDProperty()`, because whether a string is an ObjectID or not is determined by the following factors:

1. String pattern matches that of ObjectID
2. `strictObjectIDCoercion`
3. `dataType: "ObjectID"`

By line `1903`, factors 2 and 3 are not applicable, and the string looks like an ObjectID, so we simply coerce etc.

Reply:
@jannyHou is it clear to you now?

Reply:
@hacksparrow Thanks for the answer fair enough 👍